### PR TITLE
[SPARK-31962][SQL] Provide modifiedAfter and modifiedBefore options when filtering from a batch-based file data source

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

Two new options, _modifiiedBefore_  and _modifiedAfter_, is provided expecting a value in 'YYYY-MM-DDTHH:mm:ss' format.  _PartioningAwareFileIndex_ considers these options during the process of checking for files, just before considering applied _PathFilters_ such as `pathGlobFilter.`  In order to filter file results, a new PathFilter class was derived for this purpose.  General house-keeping around classes extending PathFilter was performed for neatness.  It became apparent support was needed to handle multiple potential path filters.  Logic was introduced for this purpose and the associated tests written.  

### Why are the changes needed?

When loading files from a data source, there can often times be thousands of file within a respective file path.  In many cases I've seen, we want to start loading from a folder path and ideally be able to begin loading files having modification dates past a certain point.  This would mean out of thousands of potential files, only the ones with modification dates greater than the specified timestamp would be considered.  This saves a ton of time automatically and reduces significant complexity managing this in code.

### Does this PR introduce _any_ user-facing change?

This PR introduces an option that can be used with batch-based Spark file data sources.  A documentation update was made to reflect an example and usage of the new data source option.

**Example Usages**  
_Load all CSV files modified after date:_ 
`spark.read.format("csv").option("modifiedAfter","2020-06-15T05:00:00").load()`

_Load all CSV files modified before date:_
`spark.read.format("csv").option("modifiedBefore","2020-06-15T05:00:00").load()`  

_Load all CSV files modified between two dates:_ 
`spark.read.format("csv").option("modifiedAfter","2019-01-15T05:00:00").option("modifiedBefore","2020-06-15T05:00:00").load()  
`

### How was this patch tested?

A handful of unit tests were added to support the positive, negative, and edge case code paths. 

It's also live in a handful of our Databricks dev environments.  (quoted from @cchighman)

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
cc. @maropu @gengliangwang @cloud-fan @bart-samwel @zsxwing @dongjoon-hyun @HyukjinKwon 

I've cc-ed all reviewers who left at least one review comment in original PR (#28841)

Just FYI to @cchighman as he's an author of original PR.

**Comment by HeartSaVioR** [comment]
First commit squashed the commits in origin PR. Remaining commits are my own to address my own review comments. So if you've done with the origin PR and it looked good to you, you may probably want to only look at remaining commits.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/PathFilterSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/PathFilterSuite.scala:193]: Note to reviewers: two tests were moved from FileBasedDataSourceSuite as this PR adds the specific suite for filtering option.

File: python/pyspark/sql/readwriter.py
Comment (maropu) [inline python/pyspark/sql/readwriter.py:836]: a unnecessary change?

File: python/pyspark/sql/readwriter.py
Comment (maropu) [inline python/pyspark/sql/readwriter.py:849]: ditto (I have the same comments on the changes below, too).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala:60]: `protected` -> `private`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamOptions.scala
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamOptions.scala:41]: nit: `file stream source` -> `file stream sources`?

**Review by maropu — APPROVED** [review]
I left minor comments and the other parts look fine. cc: @HyukjinKwon @dongjoon-hyun @viirya

**Comment by HeartSaVioR** [comment]
Thanks for the review, @maropu !

The origin PR has been open for months, and I only refactored a bit & fixed the doc. I'll merge this in early next week if there's no further comment.

**Comment by maropu** [comment]
> The origin PR has been open for months, and I only refactored a bit & fixed the doc. I'll merge this in early next week if there's no further comment.

Yea, it looks fine to me. Thanks for the take-over, @HeartSaVioR and thanks a lot for the valuable contribution, @cchighman !

**Comment by HeartSaVioR** [comment]
Build failure is not related. I'll let it go and check just before the merge.

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
OK. No further comments, and test passed. Merging to master.

**Comment by HeartSaVioR** [comment]
Thanks all for reviewing and merging, and thanks again @cchighman for providing great stuff!

**Comment by HyukjinKwon** [comment]
I am supportive of this change FWIW. I think it's good to have this feature.

**Comment by amadav** [comment]
Does it still need to be merged?

**Comment by HeartSaVioR** [comment]
No this is merged and will be available in new minor version, release 3.1.1.

**Pull request closed without merging**
